File: src/engine/Elevator.ts
import { Direction, ElevatorState } from "./types";
import { Request } from "./Request";

/**
 * Models a single elevator car with its complete state.
 *
 * Manages:
 *  - Current floor position
 *  - Direction of travel
 *  - Operational state
 *  - Internal queue of assigned requests
 *  - Statistics (total stops, distance traveled, energy usage)
 */
export class Elevator {
  public readonly id: number;
  public readonly name: string;
  public currentFloor: number;
  public direction: Direction;
  public state: ElevatorState;
  public requests: Request[];
  public readonly minFloor: number;
  public readonly maxFloor: number;

  // --- Statistics ---
  public totalStops: number = 0;
  public totalDistance: number = 0;
  public motorStarts: number = 0;
  public idleSteps: number = 0;

  // --- Passengers on board ---
  public passengers: Request[] = [];

  constructor(
    id: number,
    startFloor: number = 0,
    minFloor: number = 0,
    maxFloor: number = 9
  ) {
    this.id = id;
    this.name = `E${id}`;
    this.currentFloor = startFloor;
    this.direction = Direction.IDLE;
    this.state = ElevatorState.STOPPED;
    this.requests = [];
    this.minFloor = minFloor;
    this.maxFloor = maxFloor;
  }

  /** Check if this elevator has any pending work. */
  get isIdle(): boolean {
    return (
      this.requests.length === 0 &&
      this.passengers.length === 0 &&
      this.direction === Direction.IDLE
    );
  }

  /** All floors this elevator needs to visit (pick-up + drop-off). */
  get targetFloors(): number[] {
    const pickups = this.requests
      .filter((r) => !r.pickedUp)
      .map((r) => r.currentFloor);
    const dropoffs = this.passengers.map((r) => r.targetFloor);
    return [...new Set([...pickups, ...dropoffs])];
  }

  /**
   * Assign a new request to this elevator's queue.
   */
  assignRequest(request: Request): void {
    this.requests.push(request);
  }

  /**
   * Move the elevator one floor in its current direction.
   * Returns a log entry describing what happened.
   */
  step(): string[] {
    const logs: string[] = [];

    if (this.state === ElevatorState.MAINTENANCE) {
      logs.push(`${this.name}: 🔧 In maintenance`);
      return logs;
    }

    // --- Pick up passengers at current floor ---
    const pickups = this.requests.filter(
      (r) => !r.pickedUp && r.currentFloor === this.currentFloor
    );
    for (const req of pickups) {
      req.pickedUp = true;
      this.passengers.push(req);
      logs.push(
        `${this.name}: 🚶 Picked up passenger at F${this.currentFloor} → F${req.targetFloor}`
      );
      this.totalStops++;
    }

    // --- Drop off passengers at current floor ---
    const dropoffs = this.passengers.filter(
      (r) => r.targetFloor === this.currentFloor
    );
    for (const req of dropoffs) {
      req.serviced = true;
      logs.push(
        `${this.name}: ✅ Dropped off passenger at F${this.currentFloor}`
      );
      this.totalStops++;
    }
    this.passengers = this.passengers.filter(
      (r) => r.targetFloor !== this.currentFloor
    );
    // Clean serviced requests
    this.requests = this.requests.filter((r) => !r.serviced);

    // --- Determine next move ---
    if (this.requests.length === 0 && this.passengers.length === 0) {
      if (this.direction !== Direction.IDLE) {
        logs.push(`${this.name}: 💤 Now idle at F${this.currentFloor}`);
      }
      this.direction = Direction.IDLE;
      this.state = ElevatorState.STOPPED;
      this.idleSteps++;
      return logs;
    }

    // Move one floor
    const prevDirection = this.direction;
    if (this.direction === Direction.UP) {
      this.currentFloor = Math.min(this.currentFloor + 1, this.maxFloor);
      this.totalDistance++;
    } else if (this.direction === Direction.DOWN) {
      this.currentFloor = Math.max(this.currentFloor - 1, this.minFloor);
      this.totalDistance++;
    }

    this.state = ElevatorState.MOVING;

    if (
      this.direction !== Direction.IDLE &&
      prevDirection === Direction.IDLE
    ) {
      this.motorStarts++;
    }

    logs.push(
      `${this.name}: ${this.direction === Direction.UP ? "⬆️" : this.direction === Direction.DOWN ? "⬇️" : "⏸️"} Moved to F${this.currentFloor}`
    );

    return logs;
  }

  /** Snapshot of the elevator state for rendering. */
  snapshot(): ElevatorSnapshot {
    return {
      id: this.id,
      name: this.name,
      currentFloor: this.currentFloor,
      direction: this.direction,
      state: this.state,
      queueSize: this.requests.length,
      passengerCount: this.passengers.length,
      targetFloors: this.targetFloors,
      totalStops: this.totalStops,
      totalDistance: this.totalDistance,
      motorStarts: this.motorStarts,
      idleSteps: this.idleSteps,
    };
  }
}

export interface ElevatorSnapshot {
  id: number;
  name: string;
  currentFloor: number;
  direction: Direction;
  state: ElevatorState;
  queueSize: number;
  passengerCount: number;
  targetFloors: number[];
  totalStops: number;
  totalDistance: number;
  motorStarts: number;
  idleSteps: number;
}

File: elevator_backend/elevator_system/core/elevator.py
from .types import Direction, ElevatorState
from .request import Request

class Elevator:
    def __init__(self, id: int, start_floor: int = 0, min_floor: int = 0, max_floor: int = 9):
        self.id = id
        self.name = f"E{id}"
        self.current_floor = start_floor
        self.direction = Direction.IDLE
        self.state = ElevatorState.STOPPED
        self.requests = []
        self.min_floor = min_floor
        self.max_floor = max_floor

        self.total_stops = 0
        self.total_distance = 0
        self.motor_starts = 0
        self.idle_steps = 0
        self.passengers = []

    @property
    def is_idle(self) -> bool:
        return len(self.requests) == 0 and len(self.passengers) == 0 and self.direction == Direction.IDLE

    @property
    def target_floors(self) -> list[int]:
        pickups = [r.current_floor for r in self.requests if not r.picked_up]
        dropoffs = [r.target_floor for r in self.passengers]
        return list(set(pickups + dropoffs))

    def assign_request(self, request: Request):
        self.requests.append(request)

    def step(self) -> list[str]:
        logs = []

        if self.state == ElevatorState.MAINTENANCE:
            logs.append(f"{self.name}: 🔧 In maintenance")
            return logs

        # Pick up passengers
        pickups = [r for r in self.requests if not r.picked_up and r.current_floor == self.current_floor]
        for req in pickups:
            req.picked_up = True
            self.passengers.append(req)
            logs.append(f"{self.name}: 🚶 Picked up passenger at F{self.current_floor} → F{req.target_floor}")
            self.total_stops += 1

        # Drop off passengers
        dropoffs = [r for r in self.passengers if r.target_floor == self.current_floor]
        for req in dropoffs:
            req.serviced = True
            logs.append(f"{self.name}: ✅ Dropped off passenger at F{self.current_floor}")
            self.total_stops += 1
            
        self.passengers = [r for r in self.passengers if r.target_floor != self.current_floor]
        self.requests = [r for r in self.requests if not r.serviced]

        # Determine next move
        if not self.requests and not self.passengers:
            if self.direction != Direction.IDLE:
                logs.append(f"{self.name}: 💤 Now idle at F{self.current_floor}")
            self.direction = Direction.IDLE
            self.state = ElevatorState.STOPPED
            self.idle_steps += 1
            return logs

        prev_direction = self.direction
        if self.direction == Direction.UP:
            self.current_floor = min(self.current_floor + 1, self.max_floor)
            self.total_distance += 1
        elif self.direction == Direction.DOWN:
            self.current_floor = max(self.current_floor - 1, self.min_floor)
            self.total_distance += 1

        self.state = ElevatorState.MOVING

        if self.direction != Direction.IDLE and prev_direction == Direction.IDLE:
            self.motor_starts += 1

        dir_icon = "⬆️" if self.direction == Direction.UP else "⬇️" if self.direction == Direction.DOWN else "⏸️"
        logs.append(f"{self.name}: {dir_icon} Moved to F{self.current_floor}")

        return logs

    def snapshot(self) -> dict:
        return {
            "id": self.id,
            "name": self.name,
            "currentFloor": self.current_floor,
            "direction": self.direction.value,
            "state": self.state.value,
            "queueSize": len(self.requests),
            "passengerCount": len(self.passengers),
            "targetFloors": self.target_floors,
            "totalStops": self.total_stops,
            "totalDistance": self.total_distance,
            "motorStarts": self.motor_starts,
            "idleSteps": self.idle_steps,
        }
